### PR TITLE
Rewrite the tk C blitting code

### PR DESCRIPTION
**Pull request by anntzer**
... to use the standard Python C-API rather than the tk API.

We do not need to support colormodes other than 2 (i.e. RGBA)
(previously there was 0 = grayscale, 1 = RGB) so it was simpler to just
deprecate the previous function and move everything to a new private one
(matplotlib.backends._backend_tk.blit).

Followup to https://github.com/matplotlib/matplotlib/pull/10486#issuecomment-366104200.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: src/_tkagg.cpp [hunk 1]
```diff
@@ -203,9 +203,61 @@ static PyObject *_tkinit(PyObject *self, PyObject *args)
     return Py_None;
 }
 
+static PyObject *mpl_tk_blit(PyObject *self, PyObject *args) {
+    PyObject *py_photo;
+    int height, width;
+    unsigned char *data_ptr;
+    int x1, x2, y1, y2;
+    if (!PyArg_ParseTuple(args, "O(iin)(iiii)",
```
Comment (QuLogic) [inline src/_tkagg.cpp]: Add a function name here?

File: src/_tkagg.cpp [hunk 2]
```diff
@@ -203,9 +203,61 @@ static PyObject *_tkinit(PyObject *self, PyObject *args)
...
+    PyObject *py_tk, *py_interpaddr;
+    Tcl_Interp *interp;
+    if (!(py_tk = PyObject_GetAttrString(py_photo, "tk")) ||
+        !(py_interpaddr = PyObject_CallMethod(py_tk, "interpaddr", NULL)) ||
+        !(interp = (Tcl_Interp *)PyLong_AsVoidPtr(py_interpaddr))) {
+        goto exit;
+    }
+    PyObject *py_photo_name;
+    char const *photo_name;
+    Tk_PhotoHandle photo;
```
Comment (QuLogic) [inline src/_tkagg.cpp]: and these too?

File: src/_tkagg.cpp [hunk 3]
```diff
@@ -203,9 +203,61 @@ static PyObject *_tkinit(PyObject *self, PyObject *args)
...
+    int height, width;
+    unsigned char *data_ptr;
+    int x1, x2, y1, y2;
+    if (!PyArg_ParseTuple(args, "O(iin)(iiii)",
+                          &py_photo,
+                          &height, &width, &data_ptr,
+                          &x1, &x2, &y1, &y2)) {
+        goto exit;
+    }
+    PyObject *py_tk, *py_interpaddr;
```
Comment (QuLogic) [inline src/_tkagg.cpp]: Probably should all have defaults?

**Comment by anntzer** [comment]
all comments handled

File: src/_tkagg.cpp [hunk 4]
```diff
@@ -203,9 +203,62 @@ static PyObject *_tkinit(PyObject *self, PyObject *args)
...
+        // TK_FIND_PHOTO (obviously) does not set an exception on failure.
+        PyErr_SetString(PyExc_ValueError, "Failed to extract Tk_PhotoHandle");
+        goto exit;
+    }
+    block.pixelPtr = data_ptr + 4 * ((height - y2) * width + x1);
+    block.width = x2 - x1;
+    block.height = y2 - y1;
+    block.pitch = 4 * width;
+    block.pixelSize = 4;
+    block.offset[0] = 0;
```
Comment (QuLogic) [inline src/_tkagg.cpp]: Can you use this to avoid the `buf.take` copy?

Comment (anntzer) [reply src/_tkagg.cpp]: done

**Comment by tacaswell** [comment]
@mattip Can you verify that this still works with pypy?

**Comment by anntzer** [comment]
Moved some more work to Python's side, avoiding any need for refcounting at all in the C extension.

**Comment by mattip** [comment]
works fine on PyPy just like before the PR (with the nightly builds, not the latest release)

Thanks for pinging me

File: src/_tkagg.cpp [hunk 5]
```diff
@@ -203,9 +203,49 @@ static PyObject *_tkinit(PyObject *self, PyObject *args)
     return Py_None;
 }
 
+static PyObject *mpl_tk_blit(PyObject *self, PyObject *args) {
```
Comment (QuLogic) [inline src/_tkagg.cpp]: Opening function braces go on a new line.

**Comment by tacaswell** [comment]
@mattip 2.2.2 does not work on pypy?

File: src/_tkagg.cpp (hunk 5 above)
Comment (anntzer) [reply src/_tkagg.cpp]: yup, fixed

**Pull request merged**